extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -527,102 +728,107 @@ public void run()
-        if (stopRequested && !publishOnStop) {
+        if (stopRequested.get() && !publishOnStop.get()) {
           throw new InterruptedException("Stopping without publishing");
         }
 
-        status = Status.PUBLISHING;
+        status = Status.FINISHING;
       }
 
-      final TransactionalSegmentPublisher publisher = (segments, commitMetadata) -> {
-        final KafkaPartitions finalPartitions = toolbox.getObjectMapper().convertValue(
-            ((Map) commitMetadata).get(METADATA_NEXT_PARTITIONS),
-            KafkaPartitions.class
-        );
-
-        // Sanity check, we should only be publishing things that match our desired end state.
-        if (!endOffsets.equals(finalPartitions.getPartitionOffsetMap())) {
-          throw new ISE("WTF?! Driver attempted to publish invalid metadata[%s].", commitMetadata);
+      for (SequenceMetadata sequenceMetadata : sequences) {
+        if (!publishingSequences.contains(sequenceMetadata.getSequenceName())) {
+          // this is done to prevent checks in sequence specific commit supplier from failing
+          sequenceMetadata.setEndOffsets(nextOffsets);
+          sequenceMetadata.updateAssignments(nextOffsets);
+          publishingSequences.add(sequenceMetadata.getSequenceName());
+          // persist already done in finally, so directly add to publishQueue
+          publishQueue.add(sequenceMetadata);
         }
+      }
 
-        final SegmentTransactionalInsertAction action;
-
-        if (ioConfig.isUseTransaction()) {
-          action = new SegmentTransactionalInsertAction(
-              segments,
-              new KafkaDataSourceMetadata(ioConfig.getStartPartitions()),
-              new KafkaDataSourceMetadata(finalPartitions)
-          );
-        } else {
-          action = new SegmentTransactionalInsertAction(segments, null, null);
-        }
-
-        log.info("Publishing with isTransaction[%s].", ioConfig.isUseTransaction());
-
-        return toolbox.getTaskActionClient().submit(action).isSuccess();
-      };
-
-      // Supervised kafka tasks are killed by KafkaSupervisor if they are stuck during publishing segments or waiting
-      // for hand off. See KafkaSupervisorIOConfig.completionTimeout.
-      final SegmentsAndMetadata published = driver.publish(
-          publisher,
-          committerSupplier.get(),
-          sequenceNames.values()
-      ).get();
+      // add Sentinel SequenceMetadata to indicate end of all sequences
+      publishQueue.add(SequenceMetadata.getSentinelSequenceMetadata());
+      waitForPublishes.await();
+      checkAndMaybeThrowException();
 
-      final Future<SegmentsAndMetadata> handoffFuture = driver.registerHandoff(published);
-      final SegmentsAndMetadata handedOff;
+      final List<SegmentsAndMetadata> handedOffList;
       if (tuningConfig.getHandoffConditionTimeout() == 0) {
-        handedOff = handoffFuture.get();
+        handedOffList = Futures.allAsList(handOffWaitList).get();
```

Comment:
Hm, you might have a visibility concurrency issue here using an ArrayList for `handOffWaitList`, even with the sentinel (http://gee.cs.oswego.edu/dl/cpj/jmm.html).

Reply:
will fix, thanks

Reply:
changed to CopyOnWriteArrayList